src/OpenLoco/src/OpenLoco.cpp

```diff
@@ -1145,6 +1145,129 @@ namespace OpenLoco
+        }
+        template<typename T>
+        constexpr auto end(const T& item)
+        {
+            return reinterpret_cast<const char*>(&item) + sizeof(T);
+        }
+        template<typename T>
+        auto bitwise_equal(const T& lhs, const T& rhs)
+        {
+            return std::equal(begin(lhs), end(lhs), // will become constexpr with C++20
+                              begin(rhs),
+                              end(rhs));
+        }
+        template<typename T1, typename T2>
+        auto bitwise_equal(const T1& lhs, const T2& rhs)
+        {
+            return std::equal(begin(lhs), end(lhs), // will become constexpr with C++20
+                              begin(rhs),
+                              end(rhs));
+        }
+
+        template<typename T1, typename T2>
+        void bitWiseLogDivergence(const std::string entity, const T1& lhs, const T2& rhs)
+        {
+            size_t size = sizeof(T1) / sizeof(char);
+            char* array_lhs = (char*)(&lhs);
+            char* array_rhs = (char*)(&rhs);
+
+            for (size_t offset = 0; offset < size; offset++)
+            {
+                if (array_lhs[offset] != array_rhs[offset])
+                {
+                    Logging::info("DIVERGENCE");
+                    Logging::info("ENTITY: {{ {} }}", entity);
+                    Logging::info("\tOFFSET: {{ {} }}", offset);
+                    Logging::info("\t   LHS: {{ {:#x} }}", array_lhs[offset]);
+                    Logging::info("\t   RHS: {{ {:#x} }}", array_rhs[offset]);
+                }
+            }
+        }
+    }
+
+    template<typename T1, typename T2>
+    void logEntityDivergence(const std::string entity, const T1& lhs, const T2& rhs)
+    {
+        for (unsigned int offset = 0; offset < sizeof(lhs); offset++)
+            unsafe::bitWiseLogDivergence(entity + "[" + std::to_string(offset) + "]", lhs[offset], rhs[offset]);
+    }
+
+    template<typename T1, typename T2>
+    void logEntityDivergence(const std::string entity, const T1& lhs, const T2& rhs, int arraySize)
+    {
+        for (int offset = 0; offset < arraySize; offset++)
+            unsafe::bitWiseLogDivergence(entity + "[" + std::to_string(offset) + "]", lhs[offset], rhs[offset]);
+    }
+
+    template<typename T>
+    void logDivergentGameStateField(const std::string entity, int offset, const T& lhs, const T& rhs)
+    {
+        if (!unsafe::bitwise_equal(lhs, rhs))
+        {
+            Logging::info("DIVERGENCE");
+            Logging::info("ENTITY: {{ {} }}", entity);
+            Logging::info("\tOFFSET: {{ {} }}", offset);
+            Logging::info("\t   LHS: {{ {:#x} }}", lhs);
+            Logging::info("\t   RHS: {{ {:#x} }}", rhs);
+        }
+    }
+
+    void compareGameStates(const fs::path& path)
+    {
+        Logging::info("Comparing reference file {} to current GameState frame", path);
+        std::unique_ptr<S5::S5File> refFile = S5::importSave(path);
```

Comment:
the other part of the S5File that would be useful to log differences is the tile elements.